pico/services/src/main/java/io/helidon/pico/services/InterceptedMethod.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.pico.services;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.function.Function;
+
+import io.helidon.common.types.AnnotationAndValue;
+import io.helidon.common.types.TypeName;
+import io.helidon.common.types.TypedElementName;
+import io.helidon.pico.DefaultInvocationContext;
+import io.helidon.pico.Interceptor;
+import io.helidon.pico.InvocationContext;
+import io.helidon.pico.InvocationException;
+import io.helidon.pico.ServiceProvider;
+
+import jakarta.inject.Provider;
+
+/**
+ * Base class, used in {@link io.helidon.pico.Interceptor} generated code. One of these instances will be created for each
+ * intercepted method.
+ *
+ * @param <I> the intercepted type
+ * @param <V> the intercepted method return type
+ */
+public abstract class InterceptedMethod<I, V> implements Function<Object[], V> {
+    private final I impl;
+    private final InvocationContext ctx;
+
+    /**
+     * The constructor.
+     *
+     * @param interceptedImpl         the intercepted instance
+     * @param serviceProvider         the service provider for the intercepted type
+     * @param serviceTypeName         the service type name
+     * @param serviceLevelAnnotations the service level annotations
+     * @param interceptors            the interceptors for the method
+     * @param methodInfo              the method element info
+     * @param methodArgInfo           the method args element info
+     */
+    protected InterceptedMethod(I interceptedImpl,
+                                ServiceProvider<?> serviceProvider,
+                                TypeName serviceTypeName,
+                                Collection<AnnotationAndValue> serviceLevelAnnotations,
+                                Collection<Provider<Interceptor>> interceptors,
+                                TypedElementName methodInfo,
+                                TypedElementName[] methodArgInfo) {
+        this.impl = Objects.requireNonNull(interceptedImpl);
+        this.ctx =  DefaultInvocationContext.builder()
+                .serviceProvider(serviceProvider)
+                .serviceTypeName(serviceTypeName)
+                .classAnnotations(serviceLevelAnnotations)
+                .interceptors(interceptors)
+                .elementInfo(methodInfo)
+                .elementArgInfo(methodArgInfo);
+//                .build()
```

Comment:
@why is `build()` not called? You are keeping a mutable instance (general comment)

Reply:
I left it out for performance reasons only, considering the high throughput we will come to expect on this. The user will see it only as a read-only bean, not a mutable type.

Reply:
pushed